### PR TITLE
Sign in with Slack

### PR DESCRIPTION
**Pull request by granthill7**
## Pre-Requisites
- [x] Yes, I updated [Authors.md](../blob/main/Authors.md) **OR** this is not my first contribution
- [x] Yes, I included and/or modified tests to cover relevent code **OR** my change is non-technical
- [x] Yes, I wrote this code entirely myself **OR** I properly attributed these changes in [Third Party Notices](../blob/main/THIRD-PARTY-NOTICES.txt)

### CODE REVIEW

**Comment by SpencerKaiser** [comment]
@granthill7 - I had a small callout in the working agreement; please open PRs as a draft until you're ready for a formal review (which will eliminate the need for `IN PROGRESS` in the title and cut down on premature Slack notifications 😉)

File: packages/api/src/api/auth/slack.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+export const slack = async (req: Request, res: Response) => {
+  const linkBase: string =
+    'https://slack.com/openid/connect/authorize?scope=openid%20email%20profile&response_type=code';
+  // Make this an env variable
+  const redirect: string =
+    '&redirect_uri=https://1376-162-92-127-85.ngrok-free.app/api/callback/slack';
+  // Does this need to be env?
+  const client: string = `&client_id=${env.slackClientID}`;
+
+  const fullLink = linkBase + redirect + client;
```
Comment (SpencerKaiser) [inline packages/api/src/api/auth/slack.ts:13]: Can you move all this up out of the function? There's no reason to keep redeclaring
```suggestion
  const slackAuthBaseUrl: string =
    'https://slack.com/openid/connect/authorize?scope=openid%20email%20profile&response_type=code';
  const queryArgs = new URLSearchParams({
    'redirect_uri': `${env.baseUrl}/api/callback/slack`,
    'client_id': env.slackClientID,
  });

  const fullLink = `${slackAuthBaseUrl}${queryArgs.toString()}`;
```

File: packages/api/src/api/callback/index.ts
Comment (SpencerKaiser) [inline packages/api/src/api/callback/index.ts:1]: This route needs `/auth` after `/api`

File: packages/api/src/api/callback/slack.ts
Comment (SpencerKaiser) [inline packages/api/src/api/callback/slack.ts:1]: Same for this path

File: packages/api/src/api/callback/slack.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  return { email, firstName, lastName };
+};
+
+export const slack = async (req: Request, res: Response) => {
+  const myCode: string = req.query.code as string;
+  const { slackClientID, slackClientSecret } = env;
+
+  // Make this an env variable
+  const redirect: string = 'https://1376-162-92-127-85.ngrok-free.app/api/callback/slack';
```
Comment (SpencerKaiser) [inline packages/api/src/api/callback/slack.ts:23]:
```suggestion
  const redirect: string = `${env.baseUrl}/api/callback/slack`;
```

File: packages/api/src/api/index.ts [hunk 3]
```diff
@@ -12,6 +14,8 @@ if (env.nodeEnv !== 'development') api.use(enforceRateLimiting);
 
 // UNPROTECTED ROUTES
 api.use('/health', health);
+api.use('/auth', auth);
+api.use('/callback', callback);
```
Comment (SpencerKaiser) [inline packages/api/src/api/index.ts:18]: This should moved into the auth base route

File: packages/api/src/env.ts [hunk 4]
```diff
@@ -7,8 +7,11 @@ export const env = setEnv({
   required: {
     nodeEnv: 'NODE_ENV',
     port: 'PORT',
+    baseURL: 'BASE_URL',
```
Comment (SpencerKaiser) [inline packages/api/src/env.ts:10]: I _think_ this is my opinion; thoughts?
```suggestion
    baseUrl: 'BASE_URL',
```

File: packages/api/src/api/index.ts (hunk 3 above)
Comment (granthill7) [reply packages/api/src/api/index.ts:18]: So `/callback` should be moved into `api/auth/index.ts`?

Comment (SpencerKaiser) [reply packages/api/src/api/index.ts:18]: It should be registered in that file so the full url would be `/api/auth/callback/slack`

**Review by SpencerKaiser — APPROVED** [review]

**Pull request merged**